src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilderExtensions.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Razor.PooledObjects;
+
+internal static class PooledArrayBuilderExtensions
+{
+    /// <summary>
+    /// Gets a mutable reference to a <see cref="PooledArrayBuilder{T}"/> stored in a <c>using</c> variable.
+    /// </summary>
+    /// <remarks>
+    /// <para>This supporting method allows <see cref="PooledArrayBuilder{T}"/>, a non-copyable <see langword="struct"/>
+    /// implementing <see cref="IDisposable"/>, to be used with <c>using</c> statements while still allowing them to
+    /// be passed by reference in calls. The following two calls are equivalent:</para>
+    ///
+    /// <code>
+    /// using var array = PooledArrayBuilder&lt;T&gt;.Empty;
+    ///
+    /// // Using the 'Unsafe.AsRef' method
+    /// Method(ref Unsafe.AsRef(in builder));
+    ///
+    /// // Using this helper method
+    /// Method(ref builder.AsRef());
```

Comment:
I don't understand this - why cannot the `Method` take `in` (or `ref readonly`) reference instead of `ref`?

Reply:
It could, but those are really variations on `ref`. This comment is taken directly from https://sourceroslyn.io/#Microsoft.CodeAnalysis/Collections/TemporaryArrayExtensions.cs,37. FWIW, I expect that most usages will use `ref`, since it's likely that this would be passed to another method to populate it.

Reply:
> those are really variations on `ref`

Yes, but `in` wouldn't need the `AsRef()` extension method (nor the analyzer, I guess) and would still allow mutating the struct (just not the pointer), I think.

Reply:
> and would still allow mutating the struct

No, it would not allow mutating the struct. It will copy the struct at the point the mutation occurs instead of copying it as part of passing an argument. Either way, the added value will not appear in the caller's view of the collection.

Reply:
Oh, right, makes sense, thanks.

Reply:
You're treading into territory that I get easily tripped up over, but I think you working on the `ref readonly` compiler feature right? :smile:

I believe that the idea around the `AsRef()` extension method for `TemporaryArray<T>` was to enforce a particular usage pattern. @sharwell might have more insight.

Reply:
And it looks like @sharwell posted before I refreshed! :smile: